File: tests/cli/integration/back-compat-section-ownership.test.ts
/**
 * cw6/T7 back-compat regression tests.
 *
 * These tests pin AC-10 (the silent-majority invariant): existing profiles
 * laid out the v1 way — i.e. with only `.claude/CLAUDE.md` and no
 * profile-root `CLAUDE.md` — must continue to materialize EXACTLY as before
 * the section-ownership feature shipped. The CLI binary is exercised end-to-
 * end via spawn so any future regression in resolver/merge/materialize that
 * silently starts touching project-root CLAUDE.md is caught at the boundary
 * the user actually sees.
 *
 * Coverage:
 *   - BC-1: profile with ONLY `.claude/CLAUDE.md` → project-root CLAUDE.md
 *           is byte-identical before and after `use`.
 *   - BC-2: profile with ONLY profile-root CLAUDE.md → `.claude/CLAUDE.md`
 *           is NOT written (the live `.claude/` tree contains no CLAUDE.md
 *           file at all when no contributor supplies one).
 *   - BC-3: profile with BOTH → both files written independently; no
 *           cross-destination content leak (the projectRoot bytes do not
 *           appear in `.claude/CLAUDE.md`, and vice versa).
 *   - BC-4: legacy project + `init` (no profile-root CLAUDE.md anywhere)
 *           injects markers; subsequent `use` of a profile that has no
 *           profile-root contribution leaves the project-root CLAUDE.md
 *           file unchanged from init's output (section between markers
 *           remains empty / equal to init's default).
 */

import { promises as fs } from "node:fs";
import * as path from "node:path";

import { afterEach, describe, expect, it } from "vitest";

import { makeFixture, type Fixture } from "../../helpers/fixture.js";

import { ensureBuilt, runCli } from "./spawn.js";

let fx: Fixture | undefined;
afterEach(async () => {
  if (fx) await fx.cleanup();
  fx = undefined;
});

describe("cw6/T7 back-compat: section-ownership invariants (AC-10)", () => {
  // ──────────────────────────────────────────────────────────────────────
  // BC-1: Profile with ONLY `.claude/CLAUDE.md` (the v1 layout).
  // ──────────────────────────────────────────────────────────────────────
  it("BC-1: profile with only .claude/CLAUDE.md leaves project-root CLAUDE.md byte-identical (AC-10)", async () => {
    await ensureBuilt();
    fx = await makeFixture({
      profiles: {
        legacy: {
          manifest: { name: "legacy" },
          // Only a .claude/CLAUDE.md — no rootFiles.
          files: { "CLAUDE.md": "LEGACY-CLAUDE-CONTENT\n" },
        },
      },
    });

    // Hand-author a project-root CLAUDE.md the user owns. We deliberately do
    // NOT add markers — this is the "user has not opted into section
    // ownership" path. The file must remain byte-identical because no
    // contributor supplies a projectRoot source, so materialize never runs
    // the splice path (no pre-flight, no marker check).
    const rootClaudeMd = path.join(fx.projectRoot, "CLAUDE.md");
    const original = "# My project\n\nUser-authored content. No markers.\n";
    await fs.writeFile(rootClaudeMd, original);
    const originalStat = await fs.stat(rootClaudeMd);

    const r = await runCli({
      args: ["--cwd", fx.projectRoot, "use", "legacy"],
    });
    expect(r.exitCode).toBe(0);

    // The project-root CLAUDE.md is byte-identical (no marker injection
    // because we did not run init, and use does not implicitly init).
    const after = await fs.readFile(rootClaudeMd, "utf8");
    expect(after).toBe(original);

    // Mtime invariant: the file is not even opened for writing.
    const afterStat = await fs.stat(rootClaudeMd);
    expect(afterStat.mtimeMs).toBe(originalStat.mtimeMs);

    // The legacy profile's .claude/CLAUDE.md DID land in the live tree.
    expect(
      await fs.readFile(path.join(fx.projectRoot, ".claude", "CLAUDE.md"), "utf8"),
    ).toBe("LEGACY-CLAUDE-CONTENT\n");

    // State.json carries no rootClaudeMdSection (the field is null/absent).
    const state = JSON.parse(
      await fs.readFile(
        path.join(fx.projectRoot, ".claude-profiles", ".meta", "state.json"),
        "utf8",
      ),
    );
    expect(state.rootClaudeMdSection ?? null).toBeNull();
  });

  // ──────────────────────────────────────────────────────────────────────
  // BC-2: Profile with ONLY profile-root CLAUDE.md.
  // ──────────────────────────────────────────────────────────────────────
  it("BC-2: profile with only profile-root CLAUDE.md does NOT write .claude/CLAUDE.md", async () => {
    await ensureBuilt();
    // Empty fixture so `init` sees a fresh project (no pre-existing
    // .claude-profiles/ → init wouldn't refuse). Stand up the profile by
    // hand AFTER init so markers are present before `use` runs.
    fx = await makeFixture({});

    const initR = await runCli({
      args: ["--cwd", fx.projectRoot, "init", "--no-hook", "--no-seed"],
    });
    expect(initR.exitCode).toBe(0);

    // Stand up the profile: manifest + a single profile-root CLAUDE.md
    // (NO .claude/ subdir → no .claude/-destination contributions).
    const profileDir = path.join(fx.projectRoot, ".claude-profiles", "rooted");
    await fs.mkdir(profileDir, { recursive: true });
    await fs.writeFile(
      path.join(profileDir, "profile.json"),
      JSON.stringify({ name: "rooted" }, null, 2),
    );
    await fs.writeFile(path.join(profileDir, "CLAUDE.md"), "ROOT-MANAGED-BODY\n");

    // Activate the profile.
    const useR = await runCli({
      args: ["--cwd", fx.projectRoot, "use", "rooted"],
    });
    expect(useR.exitCode).toBe(0);

    // The project-root CLAUDE.md has the body spliced between markers.
    const rootContent = await fs.readFile(
      path.join(fx.projectRoot, "CLAUDE.md"),
      "utf8",
    );
    expect(rootContent).toContain("<!-- claude-profiles:v1:begin");
    expect(rootContent).toContain("<!-- claude-profiles:v1:end");
    expect(rootContent).toContain("ROOT-MANAGED-BODY");

    // CRITICAL: `.claude/CLAUDE.md` was NOT written. The live `.claude/`
    // tree contains no CLAUDE.md because no contributor supplied one for
    // that destination. (The directory itself may exist if other files
    // landed, but for this profile there are none, so the dir is empty.)
    const claudeMdLive = path.join(fx.projectRoot, ".claude", "CLAUDE.md");
    await expect(fs.access(claudeMdLive)).rejects.toThrow();

    // No content leak the other way: the projectRoot bytes are not
    // accidentally written into .claude/ either.
    let claudeDirEntries: string[] = [];
    try {
      claudeDirEntries = await fs.readdir(path.join(fx.projectRoot, ".claude"));
    } catch (err: unknown) {
      // OK if .claude/ itself doesn't exist — even safer.
      if ((err as NodeJS.ErrnoException).code !== "ENOENT") throw err;
    }
    expect(claudeDirEntries).not.toContain("CLAUDE.md");
  });

  // ──────────────────────────────────────────────────────────────────────
  // BC-3: Profile with BOTH — independent destinations, no content leak.
  // ──────────────────────────────────────────────────────────────────────
  it("BC-3: profile with both .claude/CLAUDE.md AND profile-root CLAUDE.md writes both independently (no leak)", async () => {
    await ensureBuilt();
    fx = await makeFixture({});

    const initR = await runCli({
      args: ["--cwd", fx.projectRoot, "init", "--no-hook", "--no-seed"],
    });
    expect(initR.exitCode).toBe(0);

    // Stand up profile post-init with BOTH source kinds (peer-of-
    // profile.json CLAUDE.md → projectRoot; .claude/CLAUDE.md → .claude/).
    const profileDir = path.join(fx.projectRoot, ".claude-profiles", "both");
    await fs.mkdir(path.join(profileDir, ".claude"), { recursive: true });
    await fs.writeFile(
      path.join(profileDir, "profile.json"),
      JSON.stringify({ name: "both" }, null, 2),
    );
    await fs.writeFile(path.join(profileDir, "CLAUDE.md"), "PROJECT-ROOT-BODY\n");
    await fs.writeFile(
      path.join(profileDir, ".claude", "CLAUDE.md"),
      "CLAUDE-DIR-BODY\n",
    );

    const useR = await runCli({
      args: ["--cwd", fx.projectRoot, "use", "both"],
    });
    expect(useR.exitCode).toBe(0);

    // .claude/CLAUDE.md got the .claude/-source bytes (whole-file write).
    const claudeContent = await fs.readFile(
      path.join(fx.projectRoot, ".claude", "CLAUDE.md"),
      "utf8",
    );
    expect(claudeContent).toBe("CLAUDE-DIR-BODY\n");
    // No leak: the projectRoot bytes never appear in .claude/CLAUDE.md.
    expect(claudeContent).not.toContain("PROJECT-ROOT-BODY");

    // Project-root CLAUDE.md got the profile-root bytes spliced between
    // markers (E2 concat policy applied to a single contributor).
    const rootContent = await fs.readFile(
      path.join(fx.projectRoot, "CLAUDE.md"),
      "utf8",
    );
    expect(rootContent).toContain("<!-- claude-profiles:v1:begin");
    expect(rootContent).toContain("PROJECT-ROOT-BODY");
    // No leak the other way: the .claude/ bytes never appear in root.
    expect(rootContent).not.toContain("CLAUDE-DIR-BODY");

    // State.json carries BOTH a fingerprint entry for .claude/CLAUDE.md
    // AND a rootClaudeMdSection. They are tracked independently by the
    // (path, destination) keying introduced in cw6/T3-T5.
    const state = JSON.parse(
      await fs.readFile(
        path.join(fx.projectRoot, ".claude-profiles", ".meta", "state.json"),
        "utf8",
      ),
    );
    expect(state.fingerprint.files["CLAUDE.md"]).toBeDefined();
    expect(state.rootClaudeMdSection).not.toBeNull();
    expect(state.rootClaudeMdSection.size).toBeGreaterThan(0);
  });

  // ──────────────────────────────────────────────────────────────────────
  // BC-4: Legacy project + init — markers added; later use of a profile
  //       with no projectRoot contribution leaves root file unchanged
  //       beyond init's marker output.
  // ──────────────────────────────────────────────────────────────────────
  it("BC-4: legacy project init injects markers; subsequent use of a non-root profile preserves init's output", async () => {
    await ensureBuilt();
    // Empty project — init must run cleanly. We then stand up the
    // wholly-legacy profile by hand so init's "already initialised" check
    // does not fire.
    fx = await makeFixture({});

    const initR = await runCli({
      args: ["--cwd", fx.projectRoot, "init", "--no-hook", "--no-seed"],
    });
    expect(initR.exitCode).toBe(0);

    // Stand up a wholly-legacy profile (no projectRoot CLAUDE.md anywhere).
    const profileDir = path.join(fx.projectRoot, ".claude-profiles", "legacy");
    await fs.mkdir(path.join(profileDir, ".claude"), { recursive: true });
    await fs.writeFile(
      path.join(profileDir, "profile.json"),
      JSON.stringify({ name: "legacy" }, null, 2),
    );
    await fs.writeFile(
      path.join(profileDir, ".claude", "settings.json"),
      '{"v":"legacy"}',
    );

    const rootClaudeMd = path.join(fx.projectRoot, "CLAUDE.md");
    const afterInit = await fs.readFile(rootClaudeMd, "utf8");
    // Sanity: markers present (T6 invariant).
    expect(afterInit).toContain("<!-- claude-profiles:v1:begin");
    expect(afterInit).toContain("<!-- claude-profiles:v1:end");

    // Now activate the legacy profile. Because it contributes nothing to
    // the projectRoot destination, materialize must NOT run the section
    // splice — and the project-root CLAUDE.md must be byte-identical to
    // what init produced. (The section between the markers stays empty;
    // init's output IS the final state.)
    const useR = await runCli({
      args: ["--cwd", fx.projectRoot, "use", "legacy"],
    });
    expect(useR.exitCode).toBe(0);

    const afterUse = await fs.readFile(rootClaudeMd, "utf8");
    expect(afterUse).toBe(afterInit);

    // The .claude/ side did materialize (settings.json landed). E2's
    // deep-merge canonicaliser pretty-prints settings.json on write, so we
    // assert the parsed shape rather than byte-equality against the source.
    const settings = JSON.parse(
      await fs.readFile(
        path.join(fx.projectRoot, ".claude", "settings.json"),
        "utf8",
      ),
    );
    expect(settings).toEqual({ v: "legacy" });

    // State carries no section fingerprint (no contributor → no tracking).
    const state = JSON.parse(
      await fs.readFile(
        path.join(fx.projectRoot, ".claude-profiles", ".meta", "state.json"),
        "utf8",
      ),
    );
    expect(state.rootClaudeMdSection ?? null).toBeNull();
  });
});

File: go/tests/integration/back_compat_section_ownership_test.go
// Package integration_test — IV/T4 translation of TS
// back-compat-section-ownership.test.ts.
//
// AC-10 silent-majority invariant: profiles laid out the v1 way (only
// .claude/CLAUDE.md, no profile-root CLAUDE.md) materialize EXACTLY as before
// section-ownership shipped. Asserted at the spawn boundary the user sees.
//
// Coverage:
//   BC-1: only .claude/CLAUDE.md → project-root CLAUDE.md byte-identical.
//   BC-2: only profile-root CLAUDE.md → .claude/CLAUDE.md NOT written.
//   BC-3: both → independent destinations, no cross-leak.
//   BC-4: legacy + init → markers added; later use of non-root profile
//         leaves project-root file unchanged from init's output.
package integration_test

import (
	"encoding/json"
	"errors"
	"os"
	"path/filepath"
	"strings"
	"testing"

	"github.com/htxryan/c3p/tests/integration/helpers"
)

// BC-1: profile with only .claude/CLAUDE.md leaves project-root byte-identical.
func TestBackCompat_BC1_OnlyClaudeMd_RootUntouched(t *testing.T) {
	helpers.EnsureBuilt(t)
	fx := helpers.MakeFixture(t, helpers.FixtureSpec{
		Profiles: map[string]helpers.ProfileSpec{
			"legacy": {
				Manifest: map[string]any{"name": "legacy"},
				Files:    map[string]string{"CLAUDE.md": "LEGACY-CLAUDE-CONTENT\n"},
			},
		},
	})
	rootClaudeMd := filepath.Join(fx.ProjectRoot, "CLAUDE.md")
	original := "# My project\n\nUser-authored content. No markers.\n"
	if err := os.WriteFile(rootClaudeMd, []byte(original), 0o644); err != nil {
		t.Fatalf("write root CLAUDE.md: %v", err)
	}
	originalStat, err := os.Stat(rootClaudeMd)
	if err != nil {
		t.Fatalf("stat root: %v", err)
	}

	r := mustRun(t, helpers.SpawnOptions{Args: []string{"--cwd", fx.ProjectRoot, "use", "legacy"}})
	if r.ExitCode != 0 {
		t.Fatalf("use legacy: exit=%d stderr=%q", r.ExitCode, r.Stderr)
	}

	after, err := os.ReadFile(rootClaudeMd)
	if err != nil {
		t.Fatalf("read root after: %v", err)
	}
	if string(after) != original {
		t.Errorf("root CLAUDE.md changed: want %q got %q", original, string(after))
	}
	afterStat, err := os.Stat(rootClaudeMd)
	if err != nil {
		t.Fatalf("stat root after: %v", err)
	}
	if !afterStat.ModTime().Equal(originalStat.ModTime()) {
		t.Errorf("root mtime changed: before=%v after=%v", originalStat.ModTime(), afterStat.ModTime())
	}

	live, err := os.ReadFile(filepath.Join(fx.ProjectRoot, ".claude", "CLAUDE.md"))
	if err != nil {
		t.Fatalf("read live: %v", err)
	}
	if string(live) != "LEGACY-CLAUDE-CONTENT\n" {
		t.Errorf("live CLAUDE.md want LEGACY-CLAUDE-CONTENT got %q", string(live))
	}

	stateBytes, err := os.ReadFile(filepath.Join(fx.ProjectRoot, ".claude-profiles", ".meta", "state.json"))
	if err != nil {
		t.Fatalf("read state: %v", err)
	}
	var st struct {
		RootClaudeMdSection any `json:"rootClaudeMdSection"`
	}
	if err := json.Unmarshal(stateBytes, &st); err != nil {
		t.Fatalf("parse state: %v", err)
	}
	if st.RootClaudeMdSection != nil {
		t.Errorf("rootClaudeMdSection want null/absent, got %v", st.RootClaudeMdSection)
	}
}

// BC-2: profile with only profile-root CLAUDE.md does NOT write .claude/CLAUDE.md.
func TestBackCompat_BC2_OnlyProfileRootClaudeMd_NoLiveClaudeMd(t *testing.T) {
	helpers.EnsureBuilt(t)
	fx := helpers.MakeFixture(t, helpers.FixtureSpec{})

	initR := mustRun(t, helpers.SpawnOptions{
		Args: []string{"--cwd", fx.ProjectRoot, "init", "--no-hook", "--no-seed"},
	})
	if initR.ExitCode != 0 {
		t.Fatalf("init: exit=%d stderr=%q", initR.ExitCode, initR.Stderr)
	}

	// Stand up the profile by hand AFTER init: manifest + profile-root
	// CLAUDE.md only, NO .claude/ subdir.
	profileDir := filepath.Join(fx.ProjectRoot, ".claude-profiles", "rooted")
	if err := os.MkdirAll(profileDir, 0o755); err != nil {
		t.Fatalf("mkdir profile: %v", err)
	}
	mfBytes, _ := json.MarshalIndent(map[string]any{"name": "rooted"}, "", "  ")
	if err := os.WriteFile(filepath.Join(profileDir, "profile.json"), mfBytes, 0o644); err != nil {
		t.Fatalf("write manifest: %v", err)
	}
	if err := os.WriteFile(filepath.Join(profileDir, "CLAUDE.md"), []byte("ROOT-MANAGED-BODY\n"), 0o644); err != nil {
		t.Fatalf("write profile-root CLAUDE.md: %v", err)
	}

	useR := mustRun(t, helpers.SpawnOptions{Args: []string{"--cwd", fx.ProjectRoot, "use", "rooted"}})
	if useR.ExitCode != 0 {
		t.Fatalf("use rooted: exit=%d stderr=%q", useR.ExitCode, useR.Stderr)
	}

	rootContent, err := os.ReadFile(filepath.Join(fx.ProjectRoot, "CLAUDE.md"))
	if err != nil {
		t.Fatalf("read root: %v", err)
	}
	for _, want := range []string{"<!-- c3p:v1:begin", "<!-- c3p:v1:end", "ROOT-MANAGED-BODY"} {
		if !strings.Contains(string(rootContent), want) {
			t.Errorf("root CLAUDE.md missing %q: %q", want, string(rootContent))
		}
	}

	// CRITICAL: .claude/CLAUDE.md was NOT written.
	claudeMdLive := filepath.Join(fx.ProjectRoot, ".claude", "CLAUDE.md")
	if _, err := os.Stat(claudeMdLive); err == nil {
		t.Errorf(".claude/CLAUDE.md should not exist; created at %s", claudeMdLive)
	} else if !errors.Is(err, os.ErrNotExist) {
		t.Errorf("stat .claude/CLAUDE.md: unexpected error %v", err)
	}

	// No leak: nothing under .claude/ named CLAUDE.md.
	entries, err := os.ReadDir(filepath.Join(fx.ProjectRoot, ".claude"))
	if err != nil && !errors.Is(err, os.ErrNotExist) {
		t.Fatalf("readdir .claude: %v", err)
	}
	for _, e := range entries {
		if e.Name() == "CLAUDE.md" {
			t.Errorf(".claude/CLAUDE.md present in dir listing — leak")
		}
	}
}

// BC-3: profile with both .claude/CLAUDE.md AND profile-root CLAUDE.md writes
// both independently — no content leak between destinations.
func TestBackCompat_BC3_BothDestinations_NoLeak(t *testing.T) {
	helpers.EnsureBuilt(t)
	fx := helpers.MakeFixture(t, helpers.FixtureSpec{})

	initR := mustRun(t, helpers.SpawnOptions{
		Args: []string{"--cwd", fx.ProjectRoot, "init", "--no-hook", "--no-seed"},
	})
	if initR.ExitCode != 0 {
		t.Fatalf("init: exit=%d stderr=%q", initR.ExitCode, initR.Stderr)
	}

	profileDir := filepath.Join(fx.ProjectRoot, ".claude-profiles", "both")
	if err := os.MkdirAll(filepath.Join(profileDir, ".claude"), 0o755); err != nil {
		t.Fatalf("mkdir profile/.claude: %v", err)
	}
	mfBytes, _ := json.MarshalIndent(map[string]any{"name": "both"}, "", "  ")
	if err := os.WriteFile(filepath.Join(profileDir, "profile.json"), mfBytes, 0o644); err != nil {
		t.Fatalf("write manifest: %v", err)
	}
	if err := os.WriteFile(filepath.Join(profileDir, "CLAUDE.md"), []byte("PROJECT-ROOT-BODY\n"), 0o644); err != nil {
		t.Fatalf("write profile-root CLAUDE.md: %v", err)
	}
	if err := os.WriteFile(filepath.Join(profileDir, ".claude", "CLAUDE.md"), []byte("CLAUDE-DIR-BODY\n"), 0o644); err != nil {
		t.Fatalf("write .claude/CLAUDE.md: %v", err)
	}

	useR := mustRun(t, helpers.SpawnOptions{Args: []string{"--cwd", fx.ProjectRoot, "use", "both"}})
	if useR.ExitCode != 0 {
		t.Fatalf("use both: exit=%d stderr=%q", useR.ExitCode, useR.Stderr)
	}

	claudeContent, err := os.ReadFile(filepath.Join(fx.ProjectRoot, ".claude", "CLAUDE.md"))
	if err != nil {
		t.Fatalf("read .claude/CLAUDE.md: %v", err)
	}
	if string(claudeContent) != "CLAUDE-DIR-BODY\n" {
		t.Errorf(".claude/CLAUDE.md want CLAUDE-DIR-BODY got %q", string(claudeContent))
	}
	if strings.Contains(string(claudeContent), "PROJECT-ROOT-BODY") {
		t.Errorf("leak: PROJECT-ROOT-BODY in .claude/CLAUDE.md: %q", string(claudeContent))
	}

	rootContent, err := os.ReadFile(filepath.Join(fx.ProjectRoot, "CLAUDE.md"))
	if err != nil {
		t.Fatalf("read root: %v", err)
	}
	if !strings.Contains(string(rootContent), "<!-- c3p:v1:begin") {
		t.Errorf("root missing begin marker: %q", string(rootContent))
	}
	if !strings.Contains(string(rootContent), "PROJECT-ROOT-BODY") {
		t.Errorf("root missing PROJECT-ROOT-BODY: %q", string(rootContent))
	}
	if strings.Contains(string(rootContent), "CLAUDE-DIR-BODY") {
		t.Errorf("leak: CLAUDE-DIR-BODY in root CLAUDE.md: %q", string(rootContent))
	}

	stateBytes, err := os.ReadFile(filepath.Join(fx.ProjectRoot, ".claude-profiles", ".meta", "state.json"))
	if err != nil {
		t.Fatalf("read state: %v", err)
	}
	var st struct {
		Fingerprint struct {
			Files map[string]any `json:"files"`
		} `json:"fingerprint"`
		RootClaudeMdSection *struct {
			Size int `json:"size"`
		} `json:"rootClaudeMdSection"`
	}
	if err := json.Unmarshal(stateBytes, &st); err != nil {
		t.Fatalf("parse state: %v", err)
	}
	if _, ok := st.Fingerprint.Files["CLAUDE.md"]; !ok {
		t.Errorf("state.fingerprint.files missing CLAUDE.md entry: %+v", st.Fingerprint.Files)
	}
	if st.RootClaudeMdSection == nil {
		t.Errorf("state.rootClaudeMdSection want non-null")
	} else if st.RootClaudeMdSection.Size <= 0 {
		t.Errorf("state.rootClaudeMdSection.size want >0 got %d", st.RootClaudeMdSection.Size)
	}
}

// BC-4: legacy project + init injects markers; subsequent use of a profile
// with no projectRoot contribution preserves init's output byte-for-byte.
func TestBackCompat_BC4_LegacyInitThenLegacyUse_RootPreserved(t *testing.T) {
	helpers.EnsureBuilt(t)
	fx := helpers.MakeFixture(t, helpers.FixtureSpec{})

	initR := mustRun(t, helpers.SpawnOptions{
		Args: []string{"--cwd", fx.ProjectRoot, "init", "--no-hook", "--no-seed"},
	})
	if initR.ExitCode != 0 {
		t.Fatalf("init: exit=%d stderr=%q", initR.ExitCode, initR.Stderr)
	}

	profileDir := filepath.Join(fx.ProjectRoot, ".claude-profiles", "legacy")
	if err := os.MkdirAll(filepath.Join(profileDir, ".claude"), 0o755); err != nil {
		t.Fatalf("mkdir profile/.claude: %v", err)
	}
	mfBytes, _ := json.MarshalIndent(map[string]any{"name": "legacy"}, "", "  ")
	if err := os.WriteFile(filepath.Join(profileDir, "profile.json"), mfBytes, 0o644); err != nil {
		t.Fatalf("write manifest: %v", err)
	}
	if err := os.WriteFile(filepath.Join(profileDir, ".claude", "settings.json"), []byte(`{"v":"legacy"}`), 0o644); err != nil {
		t.Fatalf("write settings: %v", err)
	}

	rootClaudeMd := filepath.Join(fx.ProjectRoot, "CLAUDE.md")
	afterInit, err := os.ReadFile(rootClaudeMd)
	if err != nil {
		t.Fatalf("read root after init: %v", err)
	}
	for _, want := range []string{"<!-- c3p:v1:begin", "<!-- c3p:v1:end"} {
		if !strings.Contains(string(afterInit), want) {
			t.Errorf("after init missing marker %q: %q", want, string(afterInit))
		}
	}

	useR := mustRun(t, helpers.SpawnOptions{Args: []string{"--cwd", fx.ProjectRoot, "use", "legacy"}})
	if useR.ExitCode != 0 {
		t.Fatalf("use legacy: exit=%d stderr=%q", useR.ExitCode, useR.Stderr)
	}

	afterUse, err := os.ReadFile(rootClaudeMd)
	if err != nil {
		t.Fatalf("read root after use: %v", err)
	}
	if string(afterUse) != string(afterInit) {
		t.Errorf("root CLAUDE.md changed during use:\nbefore=%q\nafter=%q", string(afterInit), string(afterUse))
	}

	settingsBytes, err := os.ReadFile(filepath.Join(fx.ProjectRoot, ".claude", "settings.json"))
	if err != nil {
		t.Fatalf("read settings: %v", err)
	}
	var settings map[string]any
	if err := json.Unmarshal(settingsBytes, &settings); err != nil {
		t.Fatalf("parse settings: %v", err)
	}
	if settings["v"] != "legacy" {
		t.Errorf("settings.v want legacy got %v", settings["v"])
	}

	stateBytes, err := os.ReadFile(filepath.Join(fx.ProjectRoot, ".claude-profiles", ".meta", "state.json"))
	if err != nil {
		t.Fatalf("read state: %v", err)
	}
	var st struct {
		RootClaudeMdSection any `json:"rootClaudeMdSection"`
	}
	if err := json.Unmarshal(stateBytes, &st); err != nil {
		t.Fatalf("parse state: %v", err)
	}
	if st.RootClaudeMdSection != nil {
		t.Errorf("rootClaudeMdSection want null/absent, got %v", st.RootClaudeMdSection)
	}
}
